kklib/src/os.c

```diff
@@ -246,15 +246,29 @@ kk_decl_export int kk_os_write_text_file(kk_string_t path, kk_string_t content,
 
 kk_decl_export int kk_os_read_line(kk_string_t* result, kk_context_t* ctx)
 {
+  bool eol_reached = false;
+  kk_string_t line = kk_string_empty();
   char buf[1024];
-  if (fgets(buf, 1023, stdin) == NULL) return errno;
-  buf[1023] = 0;      // ensure zero termination
-  const size_t len = strlen(buf);
-  if (len > 0 && buf[len-1] == '\n') {
-    buf[len-1] = 0;   // remove possible ending newline character
+  while (fgets(buf, sizeof buf, stdin) != NULL) {
+    kk_ssize_t len = kk_sstrlen(buf);
+    // check if we have a reached end-of-line
+    if (len > 0 && buf[len-1] == '\n') {
+      buf[--len] = '\0';  // remove ending newline character
+      eol_reached = true; // flag that we reached end-of-line
+    }
```

Comment:
I agree with your concerns. My understanding is that `fgets` will always return `NULL` if zero chars are read. So `len` should only be zero if there are`'\0'` chars in the input stream, as `strlen(buf)` will evaluate less than the chars actually read by `fgets`.

However, while thinking about this function I realised I shouldn't have used a `eol_reached` flag at all, since `fgets` also handles some cases where end-of-line is reached (e.g. `EOF` at new iteration).

I have submitted a new version where `fgets` is responsible for `EOF`s and we are responsible for `'\n'`s only. Hopefully this makes the code more descriptive and easier to reason about.